### PR TITLE
Display annotation prompt placeholder

### PR DESCRIPTION
**Pull request by tuespetre**
Addresses #3723

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DisplayMetadata.cs [hunk 1]
```diff
@@ -116,6 +116,12 @@ public class DisplayMetadata
         public int Order { get; set; } = 10000;
 
         /// <summary>
+        /// Gets or sets the placeholder.
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DisplayMetadata.cs]: Recommend stating that this is a *delegate* and not the placeholder itself. I recommend using the text from the `Description` attribute because it's the same pattern. The comment might also want to say "a delegate that gets the placeholder *text*".

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 2]
```diff
@@ -25,6 +25,9 @@ public class DefaultHtmlGenerator : IHtmlGenerator
         private const string HiddenListItem = @"<li style=""display:none""></li>";
         private static readonly MethodInfo ConvertEnumFromStringMethod =
             typeof(DefaultHtmlGenerator).GetTypeInfo().GetDeclaredMethod(nameof(ConvertEnumFromString));
+        // See: HTML5 spec, 4.10.5
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Would be nice to link to the W3C spec if possible.

Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: (To the specific section.)

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq.Expressions;
+using Microsoft.AspNet.Mvc.ModelBinding;
+using Microsoft.AspNet.Mvc.TestCommon;
+using Microsoft.AspNet.Mvc.ViewFeatures;
+using Microsoft.AspNet.Testing;
+using Xunit;
+using Microsoft.AspNet.Mvc.ModelBinding.Metadata;
```
Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs]: Sort

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperPasswordTest.cs [hunk 4]
```diff
@@ -10,6 +10,7 @@
 using Microsoft.AspNet.Mvc.ViewFeatures;
 using Microsoft.AspNet.Testing;
 using Xunit;
+using Microsoft.AspNet.Mvc.ModelBinding.Metadata;
```
Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperPasswordTest.cs]: Sort

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace Microsoft.AspNet.Mvc.Rendering
+{
+    public class HtmlHelperTextBoxTest
```
Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs]: Nice test! Would be good to also add negative tests: That is, make sure that even with a `Prompt` set, that "incompatible" input types (ones not in the list you created) do *not* get placeholder text.

**Comment by Eilon** [comment]
@tuespetre thanks for the PR! Apologies for the delay - it's holiday season here and only about 10% of the people are at the office :smile:

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs [hunk 6]
```diff
@@ -432,6 +432,20 @@ public override int Order
...
+        public override string Placeholder
+        {
+            get
+            {
+                if (DisplayMetadata.Placeholder == null)
+                {
+                    return null;
+                }
+
+                return DisplayMetadata.Placeholder();
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: nit: This can be simplified to `DisplayMetadata.Placeholder?.Invoke()` if you like.

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: Would be nice to add a comment here that this shouldn't be cached because it's localizable.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 7]
```diff
@@ -1156,12 +1159,23 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
...
             tagBuilder.MergeAttributes(htmlAttributes);
-            tagBuilder.MergeAttribute("type", GetInputTypeString(inputType));
+            tagBuilder.MergeAttribute("type", inputTypeString);
             tagBuilder.MergeAttribute("name", fullName, replaceExisting: true);
 
+            if (_placeholderInputTypes.Contains(inputTypeString))
+            {
+                modelExplorer = modelExplorer ?? ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
+
+                if (modelExplorer.Metadata.Placeholder != null)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: We should cache this in a local since it's calling into arbitrary user code.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper(model);
+
+            // Act
+            var result = helper.TextBoxFor(m => m.Property1, new { type });
+
+            // Assert
+            Assert.Equal(string.Format(expectedResult, type), HtmlContentUtilities.HtmlContentToString(result));
+        }
+
+        public class TextBoxModel
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs]: private

**Comment by tuespetre** [comment]
Thanks for the feedback! Between getting out of the holidays and getting over being sick, I am finally ready to address these points. I'll get to them as soon as I catch up on what I've missed here at my work.

**Comment by Eilon** [comment]
@tuespetre thanks, looking forward to it.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs (hunk 6 above)
Comment (tuespetre) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: When I was messing around with my own micro ORM I found that `Invoke()` was slower than a simple `()` invocation; I don't know if that's changed at all so I'll leave it 'expanded' but I will add the comment. Thanks!

**Comment by tuespetre** [comment]
My checklist:

- [ ] https://github.com/aspnet/Mvc/pull/3788#discussion_r48496142
- [ ] https://github.com/aspnet/Mvc/pull/3788#discussion_r48496163
- [ ] https://github.com/aspnet/Mvc/pull/3788#discussion_r48496229
- [ ] https://github.com/aspnet/Mvc/pull/3788#discussion_r48496239
- [ ] https://github.com/aspnet/Mvc/pull/3788#discussion_r48496277
- [ ] https://github.com/aspnet/Mvc/pull/3788#discussion_r48496380
- [ ] https://github.com/aspnet/Mvc/pull/3788#discussion_r48499078
- [ ] https://github.com/aspnet/Mvc/pull/3788#discussion_r48499131

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs (hunk 7 above)
Comment (tuespetre) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: If you mean the ExpressionMetadataProvider bit, I think I'll leave that one alone as I see there are four total such usages in this class and I'm not sure what the team would like to do with that.

**Comment by tuespetre** [comment]
Changes made and pushed up. Thanks again

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 9]
```diff
@@ -1154,12 +1157,52 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
...
+
+            if (inputType == InputType.Text || inputType == InputType.Password)
+            {
+                object suppliedTypeString;
+                if (htmlAttributes != null &&
+                    htmlAttributes.TryGetValue("type", out suppliedTypeString) &&
+                    suppliedTypeString != null &&
+                    _placeholderInputTypes.Contains(suppliedTypeString))
+                {
+                }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: @tuespetre - what goes here?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 10]
```diff
@@ -1154,12 +1157,52 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
...
+                    }
+                    else if (inputTypeString == "text")
+                    {
+                        goto case InputType.Password;
+                    }
+                    break;
+
+                case InputType.Password:
+                    modelExplorer = modelExplorer ?? ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
+                    if (modelExplorer.Metadata.Placeholder != null)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: What I meant in [this comment](https://github.com/aspnet/Mvc/pull/3788#discussion_r48499078) is that calling `.Placeholder` invokes some user code which might call back into localization, pull this out into a variable.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 11]
```diff
@@ -1154,12 +1157,52 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
...
+                        suppliedTypeString != null)
+                    {
+                        if (_placeholderInputTypes.Contains(suppliedTypeString))
+                        {
+                            goto case InputType.Password;
+                        }
+                    }
+                    else if (inputTypeString == "text")
+                    {
+                        goto case InputType.Password;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: This is a really strange way to formulate this logic. It looks like this is in the middle of a refactor and not quite finished.

 I'd expect to see something like:

```
var canIncludePlaceholder false;

switch
{
case InputType.Text:
  if (...) // conditions for text
  {
    canIncludePlaceholder = true;
  }
  break;
case InputType.Text:
  if (...) // conditions for password
  {
    canIncludePlaceholder = true;
  }
  break;
}

if (canIncludePlaceholder)
{
   // Add the placeholder
}
```

**Comment by rynowak** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs (hunk 11 above)
Comment (tuespetre) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Very true. I want to say was trying to avoid declaring a local but with all I've learned in the past week about MSIL I can now see why that was silly.

- Derek Gray

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs (hunk 10 above)
Comment (tuespetre) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Isn't that what was desired though? To not cache the result of invoking `Description()`, `DisplayName()`, and `Placeholder()`? Or should the `DefaultModelMetadata` invoke those at most one time and cache them in private fields?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs (hunk 9 above)
Comment (tuespetre) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Well, apparently I left my head there. Thank you! :fearful:

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs [hunk 12]
```diff
@@ -32,6 +32,12 @@ public class DefaultModelMetadata : ModelMetadata
         private ModelPropertyCollection _properties;
         private bool? _validateChildren;
         private ReadOnlyCollection<object> _validatorMetadata;
+        private bool _invokedDescription = false;
```
Comment (Eilon) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: BTW don't need to set any of these default values.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs [hunk 13]
```diff
@@ -32,6 +32,12 @@ public class DefaultModelMetadata : ModelMetadata
...
+        private string _cachedDescription = null;
+        private bool _invokedDisplayName = false;
+        private string _cachedDisplayName = null;
+        private bool _invokedPlaceholder = false;
+        private string _cachedPlaceholder = null;
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: None of these values should be cached. They are culture-specific and `ModelMetadata` instances have very long lives.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 14]
```diff
@@ -25,6 +25,9 @@ public class DefaultHtmlGenerator : IHtmlGenerator
         private const string HiddenListItem = @"<li style=""display:none""></li>";
         private static readonly MethodInfo ConvertEnumFromStringMethod =
             typeof(DefaultHtmlGenerator).GetTypeInfo().GetDeclaredMethod(nameof(ConvertEnumFromString));
+        // See: (http://www.w3.org/TR/html5/forms.html#the-input-element)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Minor but a blank line before this comment should make it more readable.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 15]
```diff
@@ -1159,12 +1162,47 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
...
+            var inputTypeString = GetInputTypeString(inputType);
             var tagBuilder = new TagBuilder("input");
             tagBuilder.TagRenderMode = TagRenderMode.SelfClosing;
             tagBuilder.MergeAttributes(htmlAttributes);
-            tagBuilder.MergeAttribute("type", GetInputTypeString(inputType));
+            tagBuilder.MergeAttribute("type", inputTypeString);
             tagBuilder.MergeAttribute("name", fullName, replaceExisting: true);
 
+            var tryIncludingPlaceholder = false;
+            switch (inputType)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Suggest merging this into the existing `switch` on `inputType`.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs (hunk 13 above)
Comment (tuespetre) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: Ah, ok. I wasn't aware that they have long lives. I suppose now I am just not sure of what @rynowak means about storing `Placeholder` in a local.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs (hunk 15 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Actually, the only guaranteed thing about the `inputType` is that it controls where `value` is found. Instead of all of this, do the following after the existing `switch`:
``` c#
    string typeString = tagBuilder.Attributes["type"];
    if (_placeholderInputTypes.Contains(suppliedTypeString))
    {
        modelExplorer = modelExplorer ??
            ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
        if (modelExplorer.Metadata.Placeholder != null)
        {
            tagBuilder.MergeAttribute("placeholder", modelExplorer.Metadata.Placeholder);
        }
    }
```

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+        [Theory]
+        [InlineData("text", InputWithPlaceholderAttribute)]
+        [InlineData("search", InputWithPlaceholderAttribute)]
+        [InlineData("url", InputWithPlaceholderAttribute)]
+        [InlineData("tel", InputWithPlaceholderAttribute)]
+        [InlineData("email", InputWithPlaceholderAttribute)]
+        [InlineData("number", InputWithPlaceholderAttribute)]
+        [InlineData("date", InputWithoutPlaceholderAttribute)]
+        [InlineData("color", InputWithoutPlaceholderAttribute)]
+        public void TextBoxFor_GeneratesPlaceholderAttribute_WhenDisplayAttributePromptIsSet(string type, string expectedResult)
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs]: Might be slightly more clear if split into two `[Theory]`s, one that checks output with the placeholder and one without.

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs (hunk 13 above)
Comment (tuespetre) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: Nevermind, I totally see it now. Sorry @rynowak!

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs (hunk 16 above)
Comment (tuespetre) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs]: Agreed. Will do that and make sure the `InlineData` covers all of the relevant `type` attribute values for `input` between both theories.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs (hunk 15 above)
Comment (tuespetre) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Just verified that works. Thanks @dougbu

**Comment by tuespetre** [comment]
I've completed those items as of yesterday; just pushed again to resolve the conflict

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 17]
```diff
@@ -1159,12 +1163,24 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
...
+            if (_placeholderInputTypes.Contains(suppliedTypeString))
+            {
+                modelExplorer = modelExplorer ?? ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
+                var placeholder = modelExplorer.Metadata.Placeholder;
+                if (placeholder != null)
+                {
+                    tagBuilder.MergeAttribute("placeholder", placeholder);
+                }
+            }
+
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: This is 100% better. :fireworks: :100: :fireworks:

**Comment by rynowak** [comment]
To address this comment: https://github.com/aspnet/Mvc/pull/3788#discussion_r50894328

Don't sweat stuff like this, we optimize for readability/maintainability. If there's a need to follow a specific pattern for perf, we'll let you know.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs (hunk 17 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: @dougbu is this right or is the `switch` here needed?

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs [hunk 18]
```diff
@@ -203,12 +203,12 @@ public override string Description
...
             {
-                if (DisplayMetadata.Description == null)
+                if (DisplayMetadata.Description != null)
                 {
-                    return null;
+                    return DisplayMetadata.Description();
                 }
 
-                return DisplayMetadata.Description();
+                return null;
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: We tend to leave existing style alone -- except in the case of bizarre and hard to maintain line wrapping. Why did this and `DisplayName` change at all?

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs (hunk 17 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: @rynowak no need for a `switch`.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 19]
```diff
@@ -1159,12 +1163,24 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
...
-            tagBuilder.MergeAttribute("type", GetInputTypeString(inputType));
+            tagBuilder.MergeAttribute("type", inputTypeString);
             tagBuilder.MergeAttribute("name", fullName, replaceExisting: true);
 
+            var suppliedTypeString = tagBuilder.Attributes["type"];
+            if (_placeholderInputTypes.Contains(suppliedTypeString))
+            {
+                modelExplorer = modelExplorer ?? ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
+                var placeholder = modelExplorer.Metadata.Placeholder;
+                if (placeholder != null)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: @tuespetre shouldn't this be `string.IsNullOrEmpty(placeholder)`? I.e. when is it useful to generate `placeholder=""`?

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+        [InlineData("time")]
+        [InlineData("range")]
+        [InlineData("color")]
+        [InlineData("checkbox")]
+        [InlineData("radio")]
+        [InlineData("submit")]
+        [InlineData("reset")]
+        [InlineData("button")]
+        // Skipping these two because they won't match the constant string,
+        // only because their 'value' attribute also does not get set.
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs]: Is this really a comment about `_placeholderInputTypes` in `DefaultHtmlGenerator`? If yes, should be moved.

Either way, I'm not quite sure what this sentence is saying. Could you rephrase a bit?

**Comment by dougbu** [comment]
:shipit: my comments are minor though I would like to hear from @tuespetre on https://github.com/aspnet/Mvc/pull/3788#discussion-diff-51012982.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs (hunk 18 above)
Comment (tuespetre) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: Sorry, I didn't even consider that. They changed because I was reversing the mistake I made about caching the values.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs (hunk 19 above)
Comment (tuespetre) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Good catch

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs (hunk 20 above)
Comment (tuespetre) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs]: I will amend this comment with my proper anglish skillz. :wink:

Comment (tuespetre) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs]: I take that back; I'm going to amend the tests to simply check for whether the result contains the placeholder attribute rather than comparing against a formatted constant.

Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs]: Why? These tests seem fine as is

Comment (tuespetre) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs]: @rynowak that way the tests will cover every possible `type` attribute value defined for `<input>` by the HTML5 spec.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs [hunk 21]
```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Mvc.TestCommon;
+using System.ComponentModel.DataAnnotations;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.Rendering
+{
+    public class HtmlHelperTextBoxTest
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxTest.cs]: @tuespetre not :100:% required: Please add similar test methods to `InputTagHelperTest`? It uses the same code paths through `DefaultHtmlGenerator` and passes in an `htmlAttributes` argument that includes the `type` which will really be generated. Definitely do not need to double-check _all_ of these values.

**Comment by dougbu** [comment]
Thanx @tuespetre

**Comment by rynowak** [comment]
Merged, thanks! 8b726ddc0ce208b41a07194af7052dbe14eed99f
